### DISCUSSION TITLE
How to use custom Pio files ?

### DISCUSSION THREAD

**Discussion by nordseele** [earlephilhower/arduino-pico#221] [Q&A] [answered]
Hi !
I'm trying to use a ["Quadrature encoder" PIO program](https://github.com/pmarques-dev/pico-examples/tree/master/pio/quadrature_encoder) that I found on the Raspberry Pi forum. I was looking for a way to use interrupts for my encoder like I do for my push buttons when I found this Pio solution, but I'm not really sure how to use it within Arduino-pico.

As a complete newbie with PIO, I added a couple of Includes to my sketch and then put the Pio file in the same folder, but this is probably not the right way to do because I'm getting a lot of errors from the Arduino IDE. (Of course I can post the whole output if needed)

I'm wondering if I'm supposed to compile the Pio manually and then to include the .pio.h file in my sketch or is the Arduino IDE supposed to do that when it compiles the sketch ? 

Here's the program https://github.com/pmarques-dev/pico-examples/tree/master/pio/quadrature_encoder
found in [this thread](https://www.raspberrypi.org/forums/viewtopic.php?f=145&t=305620&hilit=encoder&start=25)

Any help appreciated. 
Thanks :-)

P.S: Here's my sketch 

```
#include <Arduino.h>
#include <U8g2lib.h>
#include <Wire.h>
#include "hardware/gpio.h"
#include "hardware/pio.h"
#include "hardware/timer.h"
#include "pico/stdlib.h"
#include "quadrature_encoder.pio.h"

#define BUTTON_1 22

void button_callback(uint gpio, uint32_t events) {
  Serial.println(gpio);
}


int new_value, delta, old_value = 0;

// Base pin to connect the A phase of the encoder.
// The B phase must be connected to the next pin
const uint PIN_AB = 20;

stdio_init_all();

PIO pio = pio0;
const uint sm = 0;

uint offset = pio_add_program(pio, &quadrature_encoder_program);
quadrature_encoder_program_init(pio, sm, offset, PIN_AB, 0);

void setup() {
  gpio_init(BUTTON_1);
  gpio_set_dir(BUTTON_1, GPIO_IN);
  gpio_pull_up(BUTTON_1);
  gpio_set_irq_enabled_with_callback(BUTTON_1, 0x04, 1, button_callback);

  Serial.begin(9600);
}

void loop() {
  // note: thanks to two's complement arithmetic delta will always
  // be correct even when new_value wraps around MAXINT / MININT
  new_value = quadrature_encoder_get_count(pio, sm);
  delta = new_value - old_value;
  old_value = new_value;

  printf("position %8d, delta %6d\n", new_value, delta);
  sleep_ms(100);

}
```
P.S: Sorry if it's not the right place to ask... I'm jumping between the Raspberry Pi forum, the Arduino forum and this repository, I'm not really sure where to ask for help...

**Comment by earlephilhower** [comment, marked as answer]
Yes, you need to compile the .pio file manually into a .h, and include it in your sketch.  You can use the online compiler listed in the docs or the includes `pioasm` : https://wokwi.com/tools/pioasm
 
Don't do any stdio_init stuff, that's not compatible w/the Arduino core since the core itself manages the Serial ports.

**Reply by nordseele** [reply]
Great, I will do that !
This _pioasm_ tool looks really handy. I completely missed that part in the docs.

Thanks a lot ! :-)

**Reply by Andy2No** [reply]
> Yes, you need to compile the .pio file manually into a .h, and include it in your sketch. You can use the online compiler listed in the docs or the includes `pioasm` : https://wokwi.com/tools/pioasm

It would be good to add this to the mention of PIOASM in the https://arduino-pico.readthedocs.io/en/latest/sdk.html documentation, under PIOASM (Compiling for the PIO processors) https://arduino-pico.readthedocs.io/en/latest/sdk.html#pioasm-compiling-for-the-pio-processors

There doesn't seem to be any documentation provided on how to use the output of the assembler at https://wokwi.com/tools/pioasm or in their github repo - or if there is, it's not obvious how to find it.

**Reply by earlephilhower** [reply]
You'd just include the generated header file and do the usual stuff in your code.  It's exactly as discussed in the SDK docs, so I'm not sure there's much to document here.  Would you have a suggested update?

**Reply by Andy2No** [reply]
@earlephilhower I know very little about it, so I'm not sure I'd be a good choice to write an update.  I spent some time earlier trying to track down the information on how to do it, and eventually came to this discussion thread, via google.

Just adding what you said above, to the Read The Docs documentation, would be a good start, I think.  Ideally, there should be an entry for PIO or PIOASM on the left navigation menu  there.

Everything else I found was related to using the SDK or MBED without using the arduino-pico core, so I wasn't at all sure how to make use of it.
